Add variables for blank lines.

adiroiban opened the pull request:
Scope
====

This is the initial PR for trying to fix #732 

This does not add public configuration, but just extracts the blank lines configuration into variables.

I hope that these changes are helpful and will make the code easier to read.

I am pushing this simple PR to trigger some feedback.

To solve the blank-lines issue, I could ignore all E3XX errors and them register my own blank_lines function which would be a fork from the pycodestyle version.
But I prefer to reduce the level of forking :)


Changes
=======

For a start, it just extracts the variables into a top level dict which can be easily overwritten.

How to test
=========

Tox tests seem ok,  but I don't know how to update the other tests.

If I run pycodedstle on E30not.py it gave me errors in master:

```
$ git top
eb91b79 Ian Stapleton Cordasco 2018-02-17 Merge pull request #684 from jdufresne/async-await
$ pycodestyle testsuite/E30not.py 
testsuite/E30not.py:6:1: E302 expected 2 blank lines, found 1
testsuite/E30not.py:10:1: E302 expected 2 blank lines, found 0
testsuite/E30not.py:14:1: E302 expected 2 blank lines, found 0
testsuite/E30not.py:17:1: E302 expected 2 blank lines, found 0
testsuite/E30not.py:136:1: E302 expected 2 blank lines, found 0
testsuite/E30not.py:162:1: E302 expected 2 blank lines, found 0
(build) 
```

adiroiban commented on pycodestyle.py:
```diff
@@ -328,12 +335,17 @@ def blank_lines(logical_line, blank_lines, indent_level, line_number,
     E306: def a():\n    def b():\n        pass\n    def c():\n        pass
     """
+    top_level_lines = BLANK_LINES_CONFIG['top_level']
+    method_lines = BLANK_LINES_CONFIG['method']
```
Thanks for the review. Will update it.

I will look at updating all the code. I was not sure this will be accepted and it was a quick hack to demonstrate the direction.

--------

I was expecting that for each instance where pycodestyle asks for 2 blank line, for Twisted we will have 3 :)..and for all other insances where pycodestyle asks for 1 blank line for Twisted there will be 2... reusing the same code number.

This is how it is used now... but with a hardcore fork which is a bit hard to maintain :) https://github.com/twisted/twistedchecker/blob/master/twistedchecker/checkers/pycodestyleformat.py#L218

----------

I was thinking to just disable E3XX from pycodestyle and register a new checker with custom code but this would lead to a lot of code duplication.

This is why I am proposing this path :)

sigmavirus24 replied:
I totally understand the _why_ behind this, and agree with it. I wonder if it would be plausible to have some tests which modify these values and assert compliant code, outside of the testsuite.

adiroiban replied:
Yes. I want to write tests... no doubt about it... is just that I am confused where and how I am expected to write the tests... see testsuite/E30not.py  which fails on master.

I was thinking to create a new testsuite/test_blank_lines.py file where to write some integration tests for the blank lines code.

I am open to any other suggestion.

adiroiban commented:
btw... I saw that since 3 builds ago, travis-ci is failing with python 3.7 https://travis-ci.org/PyCQA/pycodestyle/branches

adiroiban commented:
Hi @sigmavirus24 

Please check the initial tests and let me know if this is the right direction.
I am happy to continue writing more tests for the blank_lines code.

I added a diff-cover tox environment to help with checking that changes are covered by the test.
Let me know if you would like to see codecov.io reports. I can help set that up.

I have executed flake8 and found a few issues... one was minor, regarding variable names, and I fixed that.

Travis-CI fails, due to Python2.7 changes.

```
pycodestyle.py:131: FutureWarning: Possible nested set at position 1
  EXTRANEOUS_WHITESPACE_REGEX = re.compile(r'[[({] | []}),;:]')
```

adiroiban commented:
The python2.7 was reported in #728

sigmavirus24 commented on tox.ini:
```diff
@@ -4,7 +4,7 @@
 
 [tox]
-envlist = py26, py27, py32, py33, py34, py35, py36, pypy, pypy3, jython
+envlist = py26, py27, py32, py33, py34, py35, py36, pypy, pypy3, jython, patchdiff
```
Why are you adding patchdiff here? Please, let's keep this PR focused.

sigmavirus24 commented on pycodestyle.py:
```diff
@@ -1665,8 +1682,8 @@ def parse_udiff(diff, patterns=None, parent='.'):
             rv[path] = set()
-    return dict([(os.path.join(parent, path), rows)
-                 for (path, rows) in rv.items()
+    return dict([(os.path.join(parent, child), rows)
```
Why are you changing this here?

adiroiban replied:
I run flake8 and found this error... so  I went for this drive by... is 1 line :)

I have reverted this.

sigmavirus24 commented on testsuite/support.py:
```diff
@@ -83,6 +83,25 @@ def print_results(self):
         print("Test failed." if self.total_errors else "Test passed.")
 
 
+class InMemoryReport(BaseReport):
```
:+1:

hoylemd commented on testsuite/support.py:
```diff
@@ -83,6 +83,25 @@ def print_results(self):
+    """
+
+    def __init__(self, options):
+        BaseReport.__init__(self, options)
```
Why not use `super`? python 2/3 compatability?

sigmavirus24 replied:
If BaseReport is not a new-style class I think you need to do this. I don't recall what kind of class it is, though

adiroiban replied:
I will do super :)  BaseReport is new style (class BaseReport(object):)

I just remembered that pycodestuly runs with a log of python versions...and while working on this, I was lazy and just went with this syntax. Will change to super.

Thanks:)

sigmavirus24 replied:
I think 2.6+ supports new-style classes + super so you're safe :)

adiroiban replied:
Will use super :)

@sigmavirus24 , can you please confirm whether the new tests are the type of tests that you would like to see for this change?

If you are happy with this style of testing, I can continue to write more test so that they will touch all braches.

Thanks!

sigmavirus24 replied:
Yeah, these tests look fantastic, @adiroiban :)

adiroiban commented:
I didn't had much time...  and I kind of lost interest into manual codestyles and start considering using yapf :)

The unit tests have 100% coverage for the code and I tried to import the main examples from E30.py

As a drive by I have updated the documentation about tests styles.
As a first time contributor, I was a bit confused about how to write and run the tests.

I don't like the custom framework as it make it hard to debug the code.
With unittest style, I can pinpoint a single test and inspect the isolated execution.

I think that this can be ready to merge.
